src/mock_parsec/mod.rs

```diff
@@ -0,0 +1,295 @@
+        self.update_blocks();
+        Ok(Response::new())
+    }
+
+    pub fn handle_response(
+        &mut self,
+        _src: &S::PublicId,
+        _resp: Response<T, S::PublicId>,
+    ) -> Result<(), Error> {
+        self.update_blocks();
+        Ok(())
+    }
+
+    pub fn poll(&mut self) -> Option<Block<T, S::PublicId>> {
+        state::with(self.section_hash, |state| {
+            if let Some((block, holder)) = state.blocks.get(self.first_unpolled) {
+                self.first_unpolled += 1;
+                self.observations
+                    .entry(holder.clone())
+                    .or_insert_with(ObservationInfo::new)
+                    .state = ConsensusState::Polled;
+
+                Some(block.clone())
+            } else {
+                None
+            }
+        })
+    }
+
+    #[allow(unused)]
+    pub fn can_vote(&self) -> bool {
+        unimplemented!()
+    }
+
+    #[allow(unused)]
+    pub fn have_voted_for(&self, _observation: &Observation<T, S::PublicId>) -> bool {
+        unimplemented!()
+    }
+
+    pub fn has_unconsensused_observations(&self) -> bool {
+        state::with::<T, S::PublicId, _, _>(self.section_hash, |section_state| {
+            section_state.observations.iter().any(|(holder, state)| {
+                !state.consensused()
+                    || self
+                        .observations
+                        .get(&*holder)
+                        .map(|info| info.state == ConsensusState::Unconsensused)
+                        .unwrap_or(true)
+            })
+        })
+    }
+
+    pub fn our_unpolled_observations(&self) -> impl Iterator<Item = &Observation<T, S::PublicId>> {
+        self.observations
+            .iter()
+            .filter(|(_, info)| info.our && info.state != ConsensusState::Polled)
+            .map(|(holder, _)| &**holder)
+    }
+
+    fn update_blocks(&mut self) {
+        state::with(self.section_hash, |state| {
+            while let Some((block, holder)) = state.blocks.get(self.first_unconsensused) {
+                self.handle_consensus(block.payload());
+                self.first_unconsensused += 1;
+                self.observations
+                    .entry(holder.clone())
+                    .or_insert_with(ObservationInfo::new)
+                    .set_consensused();
+            }
+        })
+    }
+
+    fn handle_consensus(&mut self, observation: &Observation<T, S::PublicId>) {
+        match *observation {
+            Observation::Add { ref peer_id, .. } => {
+                let _ = self.peer_list.insert(peer_id.clone());
+            }
+            Observation::Remove { ref peer_id, .. } => {
+                let _ = self.peer_list.remove(peer_id);
+            }
+            Observation::Accusation { ref offender, .. } => {
+                let _ = self.peer_list.remove(offender);
+            }
+            _ => (),
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(bound = "")]
```

Comment:
Is that attribute really needed everywhere? Aren't `NetworkEvent` and `PublicId` both serialisable?

Reply:
Hm. I removed it and it still compiles. I remember having issues with this before until I added that attribute. Maybe `serde` is not smart enough to figure the trait bounds somehow? Anyway, the issue seems to be gone now.